### PR TITLE
add protected commands and protected help menu items

### PR DESCRIPTION
**Pull request by technetos**
This PR adds role based protection to the commands in the form of `add_protected` and `help_protected` methods.  Both of those methods take a single command guard. 

A command guard is a function pointer in the form of `(&Args) -> Result<bool>`

These guards are used to prevent commands from being executed without authorization and to improve the generation of the help menu.  

Specifically, the help menu will only display the commands available to you as a user with a set of roles.

### CODE REVIEW

File: src/commands.rs [hunk 1]
```diff
@@ -1,11 +1,29 @@
...
 use std::{collections::HashMap, sync::Arc};
 
 const PREFIX: &'static str = "?";
 pub(crate) type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
-pub(crate) type CmdPtr = Arc<dyn for<'m> Fn(Args<'m>) -> Result<()> + Send + Sync>;
+pub(crate) type GuardFn = fn(&Args) -> Result<bool>;
+
+struct Command {
+    guard: GuardFn,
+    ptr: Box<dyn for<'m> Fn(Args<'m>) -> Result<()> + Send + Sync>,
```
Comment (tinaun) [inline src/commands.rs:15]: Should this also get a type alias?

File: src/commands.rs [hunk 2]
```diff
@@ -15,24 +33,33 @@ pub struct Args<'m> {
...
+    menu: Option<HashMap<&'static str, (&'static str, GuardFn)>>,
 }
 
 impl Commands {
     pub(crate) fn new() -> Self {
         Self {
             state_machine: StateMachine::new(),
             client: HttpClient::new(),
-            menu: HashMap::new(),
+            menu: Some(HashMap::new()),
```
Comment (tinaun) [inline src/commands.rs:46]: Why an option?

Comment (technetos) [reply src/commands.rs:46]: The ownership semantics [here][1] and [here][2] are kinda difficult to deal with when borrowing the menu HashMap.  Using an Option here, I can move the menu out of its allocation as a member of `Commands` and replace it with a None. This way I get an owned instance of the menu in main.  

[1]: https://github.com/rust-lang/discord-mods-bot/blob/089759bc5ff43af3879e0cd2633f025b7d639212/src/commands.rs#L172

[2]: https://github.com/rust-lang/discord-mods-bot/blob/089759bc5ff43af3879e0cd2633f025b7d639212/src/main.rs#L148-L150

File: src/commands.rs (hunk 1 above)
Comment (technetos) [reply src/commands.rs:15]: This is the only place where that signature is used.  :woman_shrugging: its not really worth it IMO

**Review by tinaun — APPROVED** [review]
lgtm

File: src/commands.rs (hunk 1 above)
Comment (tinaun) [reply src/commands.rs:15]: oh yeah i forgot we can't do that yet...

**Pull request merged**